Fix how Vert.x routes are identified in metrics and OpenTelemetry

gsmet opened the pull request:
Management URLs were prefixed twice when absolute: http://localhost:9000http://localhost:9000/q/health

Which was defeating the logic removing the host when collecting suppressed URIs.

Fixes #36510

gsmet committed:
Do not prefix URL when URL is already absolute

Management URLs were prefixed twice when absolute:
http://localhost:9000http://localhost:9000/q/health

Which was defeating the logic removing the host when collecting
suppressed URIs.

Fixes #36510

rquinio1A commented on extensions/vertx-http/deployment/src/main/java/io/quarkus/vertx/http/deployment/devmode/ConfiguredPathInfo.java:
Based on the package, is this code path for devmode only ? I was reproducing #36510 in production mode.

gsmet replied:
Nope, it's used when building the application itself in this particular case (and yes, it probably shouldn't!).

I'm not a big fan of how things are handled here and I might have a look at improving how these paths are handled at some point (what is extremely problematic is that we resolve the path at build time while they can only be resolved at runtime).

I'm interested in having someone test it because I actually used the debugger and didn't set up a Prometheus. my debugging session seemed in line with what people observed though.

gsmet commented:
@brunobat if you could validate it indeed fixes the whole issue when you're back from PTO, that would be awesome!

brunobat commented:
Sure, will do on Thu.

brunobat reviewed (changes requested):
Unfortunately the fix doesn't work.
I see traces for the dev UI but the application spans and the /metrics one are not collected.

gsmet commented:
Could you clarify what you see and shouldn't see or the opposite?

What I was trying to fix here are the `/health` calls being collected (and in any case, we need to get this PR in, even if the fix is incomplete).

If you have other calls to clarify, let me know and I'll have a look on Monday.

brunobat commented:
Using the otel QuickStart as an example, After the change I don't see traces for the app's `/hello` endpoint.
Also, I see no metrics for the `/metrics` endpoint, after adding the Micrometer extension. 

We probably can have a quick chat about this. I feel we need a reproducer test for this.

cescoffier reviewed (changes requested):
Just read @brunobat comment - need further investigation.

gsmet commented:
FWIW, I agree the issue is broader than this particular fix but this fix actually fixes one of the issues.

Let's say you configure the OpenTelemetry quickstart with:

```
quarkus.management.enabled=true
quarkus.otel.traces.suppress-non-application-uris=true
```

You end up with `TracerRecorder.dropNonApplicationUriTargets`:

> TracerRecorder.dropNonApplicationUriTargets: [/q/arc/removed-beans, /q/dev-ui/endpoints/*, /q/arc/beans, /q/dev-ui/io.quarkus.quarkus-arc/*, /q/health-ui*, http://localhost:9000http://localhost:9000/q/health-ui, /q/dev-ui/io.quarkus.quarkus-micrometer/*, /q/health/started, /q/dev-ui/io.quarkus.quarkus-rest-client/*, /q/arc/observers, /q/open-in-ide/:fileName/:lang/:lineNumber, /q/dev-ui/json-rpc-ws, /q/health/live, http://localhost:9000http://localhost:9000/q/health, /q/dev-ui/io.quarkus.quarkus-observability-devservices/*, /q/health/group, /q/health/well, /q/arc, /q/dev, /q/dev-ui/io.quarkus.quarkus-rest/*, /q/health/group/*, /q/health/ready, /q/dev-ui/*, /q/dev-ui, /q/_static/*, /q/dev-ui/io.quarkus.quarkus-smallrye-health/*]

As you can see, all the URLs are prefixed with `/q/`. Also you can see the `/q/health` URL being `http://localhost:9000http://localhost:9000/q/health`, which is what this PR fixes.

The thing is that if you have a look the generated spans, you will quickly understand why things are not filtered:

![Screenshot from 2025-01-09 12-54-35](https://github.com/user-attachments/assets/5beb0498-15a7-4ce1-8221-34f8ef0aba03)

As you can see, requests to `/q/dev-ui` are recorded as `/dev-ui`, requests to `/q/_static/` as `/_static` (you can't see it in this screenshot but I have the static one in another screenshot) and same for a bunch of other technical URIs.

Thus I suspect these URIs can't be properly dropped.

brunobat commented:
Me and @gsmet meet and we agree there are 3 issues at least:
* host and port should never be included in the `FrameworkEndpointsBuildItem` or at least only for Dev Mode, not in prod code. We need to understand why some URLs must be absolute. 
* The fix about the double URL that was originally removed
* Pattern marching with "bla/:filename/" are being added. 
* Spans are being reported without `/q`. Reported in the previous comment.

gsmet commented:
We agreed on me digging further and come up with a good idea of what's going on.

For me, we should include what's in this PR + a fix for the span not having the correct route path. This would be backportable.

And then take a step back and fix the reason why we are having absolute URL with host:port at build time as that's not something we want (especially since the port is set at runtime). But this is for later as I'm pretty sure whatever we do to fix that won't be backportable.

gsmet commented:
I pushed two additional commits to give an idea of where the problem is. Unfortunately, it's more complex than that as we have several routes mounted at `/dev-ui/` (such as `LocalHostOnlyFilter` and `DevUICORSFilter`) so while the patches I pushed actually fixes the issue by having `/q/` prepended, we end up with `/q/dev-ui/dev-ui/dev-ui/` - which is not exactly appealing.

I will discuss this with @cescoffier .

gsmet committed:
Identify non application routes with a name

The name is used in the metrics if defined and thus we can pass a fully
qualifier path for the route.

gsmet commented:
In the end, I think I found a smart way to fix it. Vert.x has the notion of route name that is specifically used for metrics. So when building the route for the non application routes, I push a name with the full path. The name has precedence over the path when publishing the metrics.

I'll discuss all this with @cescoffier on Monday morning but I think we are on the right path.

brunobat commented:
I'll spend some time tomorrow creating some tests to float the problem and make sure we don't have regressions in the future.

gsmet commented:
Ah yes, I wanted to ask you for that. Let me know if you have any questions!

brunobat committed:
Test for non app uris

brunobat commented:
@gsmet submitted the test. It is strange because I don't get any span from the `/dev-ui` or `/health` only from resources used by the dev-ui.
More endpoints can be added if needed.

gsmet commented:
I will have a look at the test failure tomorrow, this is odd.

gsmet committed:
Avoid duplicates in FrameworkEndpointsBuildItem

gsmet committed:
Simplify OpenTelemetrySuppressNonAppUriTest a bit

gsmet committed:
Use an HashSet contract for DropTargetsSamplerTest

It's a detail but it's the contract we want and matching will be faster.

gsmet committed:
Let TestSpanExporter when we have at least X elements included

The current condition is weird because we end up saying we want at least
1, and it only returns when there are 2.

gsmet committed:
Fix wildcard dropping in DropTargetsSampler

While /q/dev-ui/* would exclude:
- /q/dev-ui/
- /q/dev-ui/whatever
it wouldn't exclude /q/dev-ui/whatever/whenever/, which was problematic.

See additional tests.

Note that I think this should be improved with a proper matcher that
handles all our matching rules efficiently but for now, it will have to
do.

gsmet committed:
Add a SuppressNonAppUri test when management interface is enabled

brunobat reviewed (approved):
This makes sense to me. 
Thanks @gsmet

gsmet commented:
Woooot, we can now merge this!
